ddht/rpc.py

```diff
@@ -0,0 +1,253 @@
+from abc import abstractmethod
+import collections
+import io
+import json
+import logging
+import pathlib
+from typing import Any, Dict, Generic, Mapping, Optional, Tuple, TypeVar, cast
+
+from async_service import Service
+from eth_utils import ValidationError
+import trio
+
+from ddht.abc import RPCHandlerAPI, RPCRequest, RPCResponse
+from ddht.typing import JSON
+
+NEW_LINE = "\n"
+
+
+def strip_non_json_prefix(raw_request: str) -> Tuple[str, str]:
+    if raw_request and raw_request[0] != "{":
+        prefix, bracket, rest = raw_request.partition("{")
+        return prefix.strip(), bracket + rest
+    else:
+        return "", raw_request
+
+
+logger = logging.getLogger("ddht.rpc")
+decoder = json.JSONDecoder()
+
+
+async def read_json(
+    socket: trio.socket.SocketType,
+    buffer: io.StringIO,
+    decoder: json.JSONDecoder = decoder,
+) -> JSON:
+    request: JSON
+
+    while True:
+        data = await socket.recv(1024)
+        buffer.write(data.decode())
+
+        bad_prefix, raw_request = strip_non_json_prefix(buffer.getvalue())
+        if bad_prefix:
+            logger.info("Client started request with non json data: %r", bad_prefix)
+            await write_error(socket, f"Cannot parse json: {bad_prefix}")
+            continue
+
+        try:
+            request, offset = decoder.raw_decode(raw_request)
+        except json.JSONDecodeError:
```

Comment:
I'm not sure what a better alternative would be, but `JSONDecodeError` is thrown in more situations than an incomplete object. If we get truly malformed JSON this will also loop forever, disconnecting might be better UX for the other side of this connection.

Reply:
Maybe checking that All '{'s have a matching '}'? It'd be nice if there was a way to do that without writing a JSON parser :thinking:

Reply:
So we have to do it this way.  We could have read only the first part of a json string and the rest hasn't come across the socket yet.  Hence, optimistic parsing, and then continuing to try and read more from the socket and see if we can parse it again.

This could be improved to do something like enforce a maximum request payload size quite easily.... which would at least get us out of the infinite loop so I'll do that.